File: app/api/internal/cache/cache.go
// Package cache wraps Redis for both caching and a simple list-based job queue.
package cache

import (
	"context"
	"time"

	"github.com/redis/go-redis/v9"
)

type Cache struct {
	rdb *redis.Client
}

func New(url string) (*Cache, error) {
	opt, err := redis.ParseURL(url)
	if err != nil {
		return nil, err
	}
	return &Cache{rdb: redis.NewClient(opt)}, nil
}

func (c *Cache) Close() error { return c.rdb.Close() }

func (c *Cache) Ping(ctx context.Context) error { return c.rdb.Ping(ctx).Err() }

func (c *Cache) Get(ctx context.Context, key string) (string, bool, error) {
	v, err := c.rdb.Get(ctx, key).Result()
	if err == redis.Nil {
		return "", false, nil
	}
	if err != nil {
		return "", false, err
	}
	return v, true, nil
}

func (c *Cache) Set(ctx context.Context, key, value string, ttl time.Duration) error {
	return c.rdb.Set(ctx, key, value, ttl).Err()
}

func (c *Cache) Del(ctx context.Context, keys ...string) error {
	return c.rdb.Del(ctx, keys...).Err()
}

// Enqueue pushes a job payload onto a list-based queue.
func (c *Cache) Enqueue(ctx context.Context, queue, payload string) error {
	return c.rdb.LPush(ctx, queue, payload).Err()
}

// Dequeue blocks up to timeout for the next job. Returns ok=false on timeout.
func (c *Cache) Dequeue(ctx context.Context, queue string, timeout time.Duration) (string, bool, error) {
	res, err := c.rdb.BRPop(ctx, timeout, queue).Result()
	if err == redis.Nil {
		return "", false, nil
	}
	if err != nil {
		return "", false, err
	}
	if len(res) != 2 {
		return "", false, nil
	}
	return res[1], true, nil // res = [queue, value]
}

File: app/api-py/app/cache.py
"""Redis: read cache + a list-based job queue — a port of the Go app's
internal/cache/cache.go. Same keys (``steps:all``) and same queue (``dojo:jobs``),
so the Go worker and this Python worker are interchangeable consumers of the
exact same queue.
"""
from __future__ import annotations

import redis.asyncio as redis
from redis.exceptions import TimeoutError as RedisTimeoutError


class Cache:
    def __init__(self, url: str) -> None:
        # decode_responses=True -> str in/out (matches the Go string API).
        self._r = redis.from_url(url, decode_responses=True)

    async def close(self) -> None:
        await self._r.aclose()

    async def ping(self) -> None:
        await self._r.ping()

    async def get(self, key: str) -> str | None:
        """Return the cached value, or None on a miss (redis-py returns None,
        the same signal the Go code models as ok=false)."""
        return await self._r.get(key)

    async def set(self, key: str, value: str, ttl_seconds: int) -> None:
        await self._r.set(key, value, ex=ttl_seconds)

    async def delete(self, *keys: str) -> None:
        if keys:
            await self._r.delete(*keys)

    async def enqueue(self, queue: str, payload: str) -> None:
        await self._r.lpush(queue, payload)

    async def dequeue(self, queue: str, timeout_seconds: int) -> str | None:
        """Block up to timeout for the next job. Returns None when no job arrives.
        BRPOP + LPUSH = FIFO, exactly as in the Go worker. A client-side socket
        read timeout is treated the same as an empty-queue timeout — the worker
        just polls again (redis-py's async socket timeout can fire on a blocking
        BRPOP before the server-side nil arrives; either way there was no job)."""
        try:
            res = await self._r.brpop([queue], timeout=timeout_seconds)
        except RedisTimeoutError:
            return None
        if res is None:
            return None
        _, value = res  # (queue, value)
        return value
